### PR TITLE
Feat: for application with PublishVersion, use ApplicationRevision as Source of Truth for running workflow 

### PR DESCRIPTION
**Pull request by Somefive**
Signed-off-by: Somefive <yd219913@alibaba-inc.com>

### Description of your changes

1. WorkflowStepDefinition, External Workflow, External Policies are added to the ApplicationRevision.
2. WorkflowStatus is added to ApplicationRevision to record the running result of corresponding workflow. 
3. When PublishVersion not changed, ApplicationController will generate AppFile based on latest ApplicationRevision instead of Application, which means the following execution of workflow will always use the spec when the workflow started.
4. PolicyWorkload generation code is simplified.

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Add tests.

### Special notes for your reviewer

This is a fork to #3467 and CLI related codes are not included in this PR.

### CODE REVIEW

File: apis/core.oam.dev/v1beta1/applicationrevision_types.go [hunk 1]
```diff
@@ -52,34 +51,38 @@ type ApplicationRevisionSpec struct {
...
+	// Workflow records the external workflow
+	Workflow *v1alpha1.Workflow `json:"workflow,omitempty"`
+}
 
-	// ResourcesConfigMap references the ConfigMap that's generated to contain all final rendered resources.
-	ResourcesConfigMap corev1.LocalObjectReference `json:"resourcesConfigMap,omitempty"`
+// ApplicationRevisionStatus is the status of ApplicationRevision
+type ApplicationRevisionStatus struct {
+	// Succeeded records if the workflow finished running with success
+	Succeeded bool `json:"succeeded"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1beta1/applicationrevision_types.go:64]: will it be better to be a string type, indicating status as:

- deploying
- finished/succeed
- failed

Comment (Somefive) [reply apis/core.oam.dev/v1beta1/applicationrevision_types.go:64]: Actually, your idea is the initial design which was implemented in #3467. I give up this design and decided to use a simple flag for the following reason:

Comment (wonderflow) [reply apis/core.oam.dev/v1beta1/applicationrevision_types.go:64]: sure, good point

**Review by leejanee — APPROVED** [review]

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 2]
```diff
@@ -232,6 +222,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowSuspending, false)
 	case common.WorkflowStateTerminated:
 		logCtx.Info("Workflow return state=Terminated")
+		handler.UpdateApplicationRevisionStatus(logCtx, handler.latestAppRev, false, app.Status.Workflow)
```
Comment (FogDong) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:225]: Wrap this in `doWorkflowFinish`?

**Review by FogDong — APPROVED** [review]

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go (hunk 2 above)
Comment (Somefive) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:225]: This is function not always called with doWorkflowFinish.

**Review by wonderflow — APPROVED** [review]

**Review by yangsoon — APPROVED** [review]

**Pull request merged**